File: benchmarks/fibonacci/c/fibonacci.c
#include <math.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

static uint64_t fibonacci(uint32_t n) {
    switch (n) {
    case 0:
        return 0;
    case 1:
        return 1;
    default:
        return fibonacci(n - 1) + fibonacci(n - 2);
    }
}

static uint64_t fibonacci_sum(uint32_t n) {
    uint64_t result = 0;
    for (uint32_t i = 1; i < n; i++) {
        result += fibonacci(i);
    }
    return result;
}

static uint64_t monotonic_ns(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (uint64_t)ts.tv_sec * 1000000000ull + (uint64_t)ts.tv_nsec;
}

static double std_dev(const double *times, size_t len, double mean) {
    if (len < 2) {
        return 0.0;
    }
    double sum = 0.0;
    for (size_t i = 0; i < len; i++) {
        double diff = times[i] - mean;
        sum += diff * diff;
    }
    return sqrt(sum / (double)(len - 1));
}

static uint64_t bench(uint64_t run_ms, uint32_t n, int emit) {
    uint64_t run_ns = run_ms * 1000000ull;
    uint64_t elapsed_total = 0;
    uint64_t result = 0;
    size_t len = 0;
    size_t cap = 64;
    double *times = malloc(cap * sizeof(double));
    if (!times) {
        fprintf(stderr, "malloc failed\n");
        exit(1);
    }

    while (elapsed_total < run_ns) {
        uint64_t start = monotonic_ns();
        result = fibonacci_sum(n);
        uint64_t elapsed = monotonic_ns() - start;
        elapsed_total += elapsed;

        if (len == cap) {
            cap *= 2;
            double *next = realloc(times, cap * sizeof(double));
            if (!next) {
                free(times);
                fprintf(stderr, "realloc failed\n");
                exit(1);
            }
            times = next;
        }
        times[len++] = (double)elapsed / 1000000.0;
    }

    if (emit) {
        double sum = 0.0;
        double min = INFINITY;
        double max = -INFINITY;
        for (size_t i = 0; i < len; i++) {
            sum += times[i];
            if (times[i] < min) {
                min = times[i];
            }
            if (times[i] > max) {
                max = times[i];
            }
        }
        double mean = sum / (double)len;
        printf("%.6f,%.6f,%.6f,%.6f,%zu,%llu\n", mean,
               std_dev(times, len, mean), min, max, len,
               (unsigned long long)result);
    }

    free(times);
    return result;
}

int main(int argc, char **argv) {
    if (argc != 4) {
        fprintf(stderr, "Usage: %s <run_ms> <warmup_ms> <n>\n", argv[0]);
        return 1;
    }

    uint64_t run_ms = strtoull(argv[1], NULL, 10);
    uint64_t warmup_ms = strtoull(argv[2], NULL, 10);
    uint32_t n = (uint32_t)strtoul(argv[3], NULL, 10);

    if (warmup_ms > 0) {
        bench(warmup_ms, n, 0);
    }
    if (run_ms > 0) {
        bench(run_ms, n, 1);
    }
    return 0;
}

File: benchmarks/fibonacci/ruby/fibonacci.rb
#!/usr/bin/env ruby
# frozen_string_literal: true

def fibonacci(n)
  return 0 if n == 0
  return 1 if n == 1

  fibonacci(n - 1) + fibonacci(n - 2)
end

def fibonacci_sum(n)
  result = 0
  (1...n).each { |i| result += fibonacci(i) }
  result
end

def std_dev(times)
  return 0.0 if times.length < 2

  mean = times.sum / times.length.to_f
  variance = times.sum { |t| (t - mean) * (t - mean) } / (times.length - 1).to_f
  Math.sqrt(variance)
end

def bench(run_ms, n)
  run_ns = run_ms * 1_000_000
  elapsed_total = 0
  times = []
  result = nil

  while elapsed_total < run_ns
    start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
    result = fibonacci_sum(n)
    elapsed = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - start
    elapsed_total += elapsed
    times << elapsed / 1_000_000.0
  end

  [times, result]
end

if ARGV.length != 3
  warn "Usage: #{$PROGRAM_NAME} <run_ms> <warmup_ms> <n>"
  exit 1
end

run_ms = ARGV[0].to_i
warmup_ms = ARGV[1].to_i
n = ARGV[2].to_i

bench(warmup_ms, n) if warmup_ms.positive?

if run_ms.positive?
  times, result = bench(run_ms, n)
  mean = times.sum / times.length.to_f
  puts format('%.6f,%.6f,%.6f,%.6f,%d,%d',
              mean, std_dev(times), times.min, times.max, times.length, result)
end
